303-add-XCUTestTarged-initialtests

isabelrios opened the pull request:
Resolves #303 

This PR adds the basic environment to have XCUITests running. In order to achieve that the main additions are:
-MappaMundi
-LockboxXCUITest target
-Added a Testsuite containing one test, LockboxXCUITests
-LockboxScreengraph and BaseTestCase to support writing more easy an efficient tests

- There are a few project files that change with different ids that I'm not very sure how to change so that the main project file is not modified. Hope I have not messed anything up with it.

- Since the app is not behaving consistently while going to background and relaunching it, there is only one test checking multiple scenarios. That will change once the behavior is stable and we can define a good tearDown method between the different tests. 
- There are some sleep waits that will be removed in the future too.
- The LockboxScreenGraph is a first draft of the app screens and transitions, it has to be improved (the back actions are not defined yet and more actions and transitions need to be added) but at least having a basic structure is a way to start defining the XCUITest environment.
- The idea would be to run the tests on BB once it is consistently passing so I have not configured anything related to BB yet.

Please let me know your comments, thoughts, all your feedback is more than welcome!
Thanks!

isabelrios commented:
@devinreams I added a few more tests but two of them are failing now. It's strange but using the same testing account as before, the login entry that we have saved does not appear...It appears correctly on other device, but not in all of them, I have tried on different simulators. Is it due to any problem already tracked? or should I file a new issue? 
Thanks!

devinreams commented:
>the login entry that we have saved does not appear...It appears correctly on other device, but not in all of them, I have tried on different simulators. Is it due to any problem already tracked? or should I file a new issue?

That's interesting. To be clear I think you're saying:

1. this new test FxA account has 1 entry that the tests are expecting
2. when tests run it signs in successfully but does not see that 1 entry

And this test FxA account does not require/expect a "confirmation email" because it was created through FxAUtils. Is that right?

isabelrios commented:
>   1. this new test FxA account has 1 entry that the tests are expecting
Well, is the old FxA account I was using, but yes, it has one entry I am able to see in one device only. It was working in all devices, the tests passing but don't know why it stopped showing the login entry.
 >  2. when tests run it signs in successfully but does not see that 1 entry
Yes, that's right. But it is not only a problem when running the tests. Using that account manually on sims or physical devices shows the same empty logins list.

> And this test FxA account does not require/expect a "confirmation email" because it was created through FxAUtils. Is that right?
Yes, that's right.
You will see on the tests this account, I keep using it but now to check the screen when there are no entries.

isabelrios commented:
Tests are passing now when there is a regular run on BB:
https://dashboard.buddybuild.com/apps/5a0ddb736e19370001034f85/build/5b085c9d71ab6f00014bf97f#tests

But if for some reason there is a retry and build starts again, then they fail. This will be solved once we define a correct tear down that leave the app after the execution in a clean state

jhugman commented on LockboxXCUITests/LockBoxScreenGraph.swift:
```diff
@@ -0,0 +1,159 @@
+    map.addScreenState(WelcomeScreen) { screenState in
+        // Adding temporary a sleep here because now Get Start buttons appears always when relaunching the app, bug or design??
+        sleep(3)
+        if (app.buttons["Get Started"].exists) {
```
This will run when the graph is constructed, not when the test navigator is at the Welcome Screen; so this exists test will always be false.

isabelrios replied:
Okay, I see...

jhugman commented on LockboxXCUITests/LockBoxScreenGraph.swift:
```diff
@@ -0,0 +1,159 @@
+            }
+        }
+
+        screenState.dismissOnUse = true
```
This screen isn't dismissed when the user takes an action.

jhugman commented on LockboxXCUITests/LockBoxScreenGraph.swift:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    map.addScreenState(LockedScreen) { screenState in
+        screenState.backAction = navigationControllerBackAction
```
If the locked screen has a back action, then this would be a bug. 

I suggest you add a couple of properties to the user state.

To unlock, you need to be using biometrics or a passcode.

isabelrios replied:
There is not a back action, no bug there, I should not have added that back action there, just added the LockedScreen to have that for the future and copy/paste from another screen state

isabelrios commented on LockboxXCUITests/LockBoxScreenGraph.swift:
```diff
@@ -0,0 +1,159 @@
+let AccountSettingsMenu = "AccountSettingsMenu"
+let AutolockSettingsMenu = "AutolockSettingsMenu"
+
+let SortEntriesMenu = "SortEntriesMenu"
```
👍

isabelrios commented on LockboxXCUITests/LockBoxScreenGraph.swift:
```diff
@@ -0,0 +1,159 @@
+        // Adding temporary a sleep here because now Get Start buttons appears always when relaunching the app, bug or design??
+        sleep(3)
+        if (app.buttons["Get Started"].exists) {
+            screenState.tap(app.buttons["Get Started"], to: FxASigninScreen)
```
Done, I added an accessibilityId for that button, hope it is the correct way.

isabelrios commented on LockboxXCUITests/LockBoxScreenGraph.swift:
```diff
@@ -0,0 +1,159 @@
+        app.navigationBars.element(boundBy: 0).buttons.element(boundBy: 0).tap()
+    }
+
+    let cancelBackAction = {
```
Yes, I have to work better on the back/cancel actions. Just 'drafted' those actions for taking them into account for the future.

isabelrios commented on LockboxXCUITests/LockBoxScreenGraph.swift:
```diff
@@ -0,0 +1,159 @@
+    map.addScreenState(LockboxMainPage) { screenState in
+        screenState.tap(app.buttons["preferences"], to: SettingsMenu)
+
+        screenState.gesture(forAction: Action.ChangeEntriesOrder, transitionTo: SortEntriesMenu) { userState in
```
You are right, it should be a gesture, changed now.

isabelrios commented on LockboxXCUITests/LockBoxScreenGraph.swift:
```diff
@@ -0,0 +1,159 @@
+
+    map.addScreenState(SortEntriesMenu) { screenState in
+        //screenState.backAction = cancelBackAction
+        screenState.noop(to: LockboxMainPage)
```
👍

isabelrios commented:
Thanks @jhugman for the review! I know you are very busy with the bug fixes so I really appreciate your time reviewing this. Sorry for some of the issues you commented, my intention was to have a draft of the graph (so all the back actions are not completely defined and so the screens) and start building from there. I tried to fix all your comments, so hopefully it looks better now.

isabelrios commented on LockboxXCUITests/LockBoxScreenGraph.swift:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    map.addScreenState(Screen.LockedScreen) { screenState in
+        //screenState.backAction = navigationControllerBackAction
```
Honestly I don't know how to manage this screen :( . I was able to lock/unlock the device within the tests but not adding here an action. It is needed something like:
```screenState.gesture(forAction: Action.UnlockDevice, transitionTo: Screen.LockboxMainPage)
            let springboard = XCUIApplication(bundleIdentifier: "com.apple.springboard")
            let passcodeInput = springboard.secureTextFields["Passcode field"]
            passcodeInput.tap()
            passcodeInput.typeText("0000\r")```

It fails becasue the scureTextField is not found, also, to me, that does not look like a part of the app to be defined on the graph...am I wrong? 
For the time being I have removed that screen and added a test for testing it until I know how to manage it better...

isabelrios commented on LockboxXCUITests/LockBoxScreenGraph.swift:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    map.addScreenState(Screen.WelcomeScreen) { screenState in
+            screenState.tap(app.buttons["getStarted.button"], to: Screen.FxASigninScreen)
```
I added two more, sorry that I didn't do at the same time than the first one, wasn't sure that the way I did it was correct. For the rest, I'm trying to get them but looks like I have to add them in a different way and have not found it so that they work yet.

isabelrios commented:
@jhugman I think I addressed your comments except for having all `accessibilityIdentifier`s. I added some of them* but for cases like the options in Settings Menu (i.e: `app.tables.cells.staticTexts["Open Websites in"]` I don't know how to do it...:( Could we complete the addition of all acc.Ids in a follow up? 

*Acc.Ids added so far: 
**Welcome screen:**
getStarted.button

**Entry List**
settings.button
sorting.button

**Settings List**
sendUsageData.switch
lockNow.button

**Account Setting**
userName.label
learnEditEntry.button

**Entry detailed view**
reveal.button

I think we should reopen issue #453 or open a new one to track this so that there are acc.IDs for all necessary items.
Thanks!

devinreams commented:
>I think we should reopen issue #453 or open a new one to track this so that there are acc.IDs for all necessary items.

Agreed, more accessibility identifiers to be added at #471

isabelrios commented:
> Agreed, more accessibility identifiers to be added at #471

Thank you! I will update that issue with the latest identifiers I added in this PR.

devinreams commented:
@isabelrios are you ready for full re-review of this pull request (is it no longer "work in progress" to you?) 😁

isabelrios commented:
> @isabelrios are you ready for full re-review of this pull request (is it no longer "work in progress" to you?) 

I addressed almost all comments from previous review and was waiting for feedback about that. Haven't touched this for some days and now looks like I need to fix the tests, after the last rebase (thanks! I was about to do it :)) the tests are failing, they will need some changes after latest updates on master. Once I fix them I will remove the WIP so that it can be fully reviewed. Thanks!

isabelrios commented:
Removing WIP, test are passing.
Obviously there is room for improvement and more tests/checks can be added. But the idea would be to have a "basic" test suite running XCUITests and improve it in terms of tests and so step by step in future PRs having in mind what is needed as the app evolves.
Please let me know your comments and thoughts.
Thanks!

devinreams commented:
>Removing WIP, test are passing.

Awesome, thanks for rebasing and getting that squared away!

And I completely agree with the immediate vs. future approach.

@jhugman and @sashei can you two please help re-review this so we can have these tests in place going forward as we are nearly a “public facing” release and want to keep quality high and potential breakage low. 😁

sashei commented:
Oops, commented on the wrong PR...

@isabelrios it looks like this is failing because of the issue with the test account, did you ever get an answer about what we should do for those cases? I'd like to use something similar when automating screenshots.

isabelrios commented:
> it looks like this is failing because of the issue with the test account, did you ever get an answer about what we should do for those cases? I'd like to use something similar when automating screenshots.

@sashei The failure was due to the new onboarding screen, that was not added to the test yet but it is there now and tests are green again.
For verifing the account so that we can have the tests always running I'm working on a solution using the [REST service](https://restmail.net/) available for @restmail.net accounts. I don't have it fully working yet on the tests but hopefully it will be soon. I guess that will be useful for the screenshots tests too...we can talk about that if/when you want :)

sashei commented:
🎉
